_posts/2024-12-12-Introducing-BMO-E2E.md

```diff
@@ -0,0 +1,107 @@
+minikube, some VMs and a baseboard management controller (BMC) emulator. The
+goal was simple: do the minimum required to simulate a baremetal lab. From this,
+it was quite easy to build a test module that was responsible for deploying the
+necessary controllers and running some tests.
+
+Notice the separation of concerns here! The test module expects a baremetal lab
+environment to be already existing and the script that sets up the environment
+is not involved in anyway with the tests or deployment of the controllers. This
+design is deliberate, with a clear goal that the test module should be useful
+across multiple environments. It should be possible to run the test suite
+against real baremetal labs with multiple different configurations. I am hoping
+that we will get a chance next year to try it for real in a baremetal lab.
+
+## How does it work?
+
+The flexibility of the end to end module is possible through a configuration
+file. It can be used to configure everything from the image URL and checksum to
+the timeout limits. Since Ironic can be deployed in many different ways, it was
+also necessary to make this flexible. The user can optionally set up Ironic
+before the test, or provide a kustomization that will be applied automatically.
+A separate configuration file declares the BMCs that should be used in the
+tests.
+
+The [configuration that we use in
+CI](https://github.com/metal3-io/baremetal-operator/tree/main/test/e2e/config)
+shows how these files look like. As a proof of concept for the flexibility of
+the tests, it can be noted that we already have two different configurations.
+One for running the tests with Ironic and one for running them with BMO in
+fixture mode. The first is the "normal" mode, the latter means that BMO does not
+communicate with Ironic at all, it just pretends. While that obviously isn't
+useful for any thorough tests, it still provides a quick and light weight test
+suite, and ensures that we do not get too attached to one particular
+configuration.
+
+The test suite itself is made with ginkgo and gomega. Instead of building a long
+chain of checks and scenarios we have attempted to do small, isolated tests.
+This makes it possible to run multiple in parallel and shorten the test suite
+duration, as well as easily identify where exactly errors occur. In order to
+accomplish this, we make heavy use of the [status
+annotation](https://book.metal3.io/bmo/status_annotation) so that we can skip
+inspection when possible.
+
+## Where are we today?
+
+It is already several months since we switched over to the BMO e2e test suite as
+the primary, and only required tests for pull requests in the BMO repository.
+The end to end test suite covers more than the metal3-dev-env and CAPM3 based
+tests from BMO perspective. That does not mean that we are done though. At the
+time of writing, there are several github issues for improving and extending the
```

Comment:
```suggestion
time of writing, there are several GitHub issues for improving and extending the
```

Should we also add a link to issue search to these issues?

Reply:
I was thinking about it but we do not have a simple way of filtering for them. There is no label and the project we have is private. Perhaps we can make it public though?

Reply:
Maybe just link them one by one then? Like "Several (#xxx, #yyy, #zzz)" style?

Reply:
I could link this if we make it public: https://github.com/orgs/metal3-io/projects/5/views/2
I don't have privileges to make it public though.

Reply:
@kashifest @Rozzii can we make it public, I see no reason why not?

Reply:
I have approved it yesterday, it can go public.

Reply:
@Rozzii no, the board is not public. It cannot be viewed by non-members. That what we're discussing, not just publishing the link.

Reply:
Okay I didn't read all the messages properly I thought the post has to go public, sorry. 
I have changed the board to public, I don't see any issue with the board being public.

Reply:
Thanks, I will update the post with a link to it :slightly_smiling_face: